lightautoml/transformers/groupby.py

```diff
@@ -0,0 +1,205 @@
+"""GroupBy (categorical/numerical) features transformer."""
+
+from typing import List
+from typing import Optional
+from typing import Union
+
+import numpy as np
+
+from scipy.stats import mode as get_mode
+
+from ..dataset.base import LAMLDataset
+from ..dataset.roles import NumericRole
+from .base import LAMLTransformer
+
+
+_transform_types_numeric = ["delta_median", "delta_mean", "min", "max", "std"]
+_transform_types_categorical = ["mode", "is_mode"]
+
+
+class GroupByTransformer(LAMLTransformer):
+    """Transformer, that calculates groupby features.
+
+    Types of group_by transformations:
+        - Numerical features:
+            - delta_median: Difference with group mode.
+            - delta_mean: Difference with group median.
+            - min: Group min.
+            - max: Group max.
+            - std: Group std.
+        - Categorical features:
+            - mode: Group mode.
+            - is_mode: Is current value equal to group mode.
+
+    Attributes:
+        features list(str): generated features names.
+
+    """
+
+    _fname_prefix = "grb"
+
+    @property
+    def features(self):
+        """Features list."""
+        return self._features
+
+    def __init__(
+        self,
+        group_col: str,
+        numeric_cols: Optional[List[str]] = None,
+        categorical_cols: Optional[List[str]] = None,
+        used_transforms: Optional[List[str]] = None,
+    ):
+        """Initialize transformer.
+
+        Args:
+            group_col: Name of categorical variable for grouping.
+            numeric_cols: List of numeric variables to calculate groupby with respect to the 'group_column'.
+            categorical_cols: List of categorical variables to calculate groupby with respect to the 'group_column'.
+            used_transforms: List of used transformation types, for example ["std", "mode", "delta_mean"].
+                             If not specified, all available transformations are used.
+        """
+        # assert set(used_transforms).issubset(_transform_types_numeric + _transform_types_categorical), \
+        # f"Only these transformation types supported: {_transform_types_numeric + _transform_types_categorical}"
+
+        super().__init__()
+        self.group_col = group_col
+        self.numeric_cols = numeric_cols
+        self.categorical_cols = categorical_cols
+        self._feat_idx = self._set_feature_indices()
+        self.used_transforms = (
+            used_transforms if used_transforms else _transform_types_numeric + _transform_types_categorical
+        )
+        self.numeric_transforms = [t for t in self.used_transforms if t in _transform_types_numeric]
+        self.categorical_transforms = [t for t in self.used_transforms if t in _transform_types_categorical]
+
+    def _set_feature_indices(self):
+        feat_idx = dict()
+        feat_idx[self.group_col] = 0
```

Comment:
The thing is that we don't have the original feature names in there, because they are affected by different encoders. So that, hardcoding idx is the only way to link them with original names. To fix this we need some general improvements: save original feature names through one-to-one transformations.